module/zcommon/zfs_fletcher_sse.c

```diff
@@ -0,0 +1,205 @@
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#if defined(HAVE_SSE2)
+
+#include <linux/simd_x86.h>
+#include <sys/spa_checksum.h>
+#include <zfs_fletcher.h>
+
+struct zfs_fletcher_sse_array {
+	uint64_t v[2] __attribute__((aligned(16)));
+};
+
+static void
+fletcher_4_sse2_init(zio_cksum_t *zcp)
+{
+	kfpu_begin();
+
+	/* clear sse registers */
+	asm volatile("pxor %xmm0, %xmm0");
+	asm volatile("pxor %xmm1, %xmm1");
+	asm volatile("pxor %xmm2, %xmm2");
+	asm volatile("pxor %xmm3, %xmm3");
+}
+
+static void
+fletcher_4_sse2_fini(zio_cksum_t *zcp)
+{
+	struct zfs_fletcher_sse_array a, b, c, d;
+	uint64_t A, B, C, D;
+
+	asm volatile("movdqa %%xmm0, %0":"=m" (a.v));
+	asm volatile("movdqa %%xmm1, %0":"=m" (b.v));
+	asm volatile("psllq $0x2, %xmm2");
+	asm volatile("movdqa %%xmm2, %0":"=m" (c.v));
+	asm volatile("psllq $0x3, %xmm3");
+	asm volatile("movdqa %%xmm3, %0":"=m" (d.v));
+
+	kfpu_end();
+
+	/*
+	 * The mixing matrix for checksum calculation is:
+	 * a = a0 + a1
+	 * b = 2b0 + 2b1 - a1
+	 * c = 4c0 - b0 + 4c1 -3b1
+	 * d = 8d0 - 4c0 + 8d1 - 8c1 + b1;
+	 *
+	 * c and d are multiplied by 4 and 8, respectively,
+	 * before spilling the vectors out to memory.
+	 */
+	A = a.v[0] + a.v[1];
+	B = 2*b.v[0] + 2*b.v[1] - a.v[1];
+	C = c.v[0] - b.v[0] + c.v[1] - 3*b.v[1];
+	D = d.v[0] - c.v[0] + d.v[1] - 2*c.v[1] + b.v[1];
+
+	ZIO_SET_CHECKSUM(zcp, A, B, C, D);
+}
+
+static void
+fletcher_4_sse2(const void *buf, uint64_t size, zio_cksum_t *unused)
+{
+	const uint64_t *ip = buf;
+	const uint64_t *ipend = (uint64_t *)((uint8_t *)ip + size);
+
+	asm volatile("pxor %xmm4, %xmm4");
+
+	for (; ip < ipend; ip += 2) {
+		asm volatile("movdqu %0, %%xmm5" :: "m"(*ip));
+		asm volatile("movdqa %xmm5, %xmm6");
+		asm volatile("punpckldq %xmm4, %xmm5");
+		asm volatile("punpckhdq %xmm4, %xmm6");
+		asm volatile("paddq %xmm5, %xmm0");
+		asm volatile("paddq %xmm0, %xmm1");
+		asm volatile("paddq %xmm1, %xmm2");
+		asm volatile("paddq %xmm2, %xmm3");
+		asm volatile("paddq %xmm6, %xmm0");
+		asm volatile("paddq %xmm0, %xmm1");
+		asm volatile("paddq %xmm1, %xmm2");
+		asm volatile("paddq %xmm2, %xmm3");
+	}
+}
+
+static void
+fletcher_4_sse2_byteswap(const void *buf, uint64_t size, zio_cksum_t *unused)
+{
+	const uint32_t *ip = buf;
+	const uint32_t *ipend = (uint32_t *)((uint8_t *)ip + size);
+
+	for (; ip < ipend; ip += 2) {
+		uint32_t scratch;
+
+		asm volatile("bswapl %0" : "=r"(scratch) : "0"(*ip));
```

Comment:
`bswap` only works on a reg, so having this one-liner might be confusing.
ZFS already has `BSWAP_32()` macros, that compiler will turn into correct instruction. IMO, this would read more clearly as: `scratch = BSWAP_32(ip[0]);`